lib/block_view_validator_test.go

```diff
@@ -144,11 +151,130 @@ func _testValidatorRegistration(t *testing.T, flushToDB bool) {
+	}
+	{
+		// RuleErrorValidatorInvalidVotingSignatureBlockHeight: too high
+		votingPublicKey, votingSignature := _generateVotingPublicKeyAndSignature(t, m0PkBytes, blockHeight)
+		registerMetadata = &RegisterAsValidatorMetadata{
+			Domains:                    [][]byte{[]byte("https://example.com")},
+			VotingPublicKey:            votingPublicKey,
+			VotingPublicKeySignature:   votingSignature,
+			VotingSignatureBlockHeight: blockHeight + params.ValidatorVotingSignatureBlockHeightWindow + 1,
+		}
+		_, err = _submitRegisterAsValidatorTxn(
+			testMeta, m0Pub, m0Priv, registerMetadata, nil, flushToDB,
+		)
+		require.Error(t, err)
+		require.Contains(t, err.Error(), RuleErrorValidatorInvalidVotingSignatureBlockHeight)
+	}
+	{
+		// RuleErrorValidatorMissingVotingPublicKeySignature
+		votingPublicKey, _ := _generateVotingPublicKeyAndSignature(t, m0PkBytes, blockHeight)
+		registerMetadata = &RegisterAsValidatorMetadata{
+			Domains:                    [][]byte{[]byte("https://example.com")},
+			VotingPublicKey:            votingPublicKey,
+			VotingSignatureBlockHeight: blockHeight,
+		}
+		_, err = _submitRegisterAsValidatorTxn(
+			testMeta, m0Pub, m0Priv, registerMetadata, nil, flushToDB,
+		)
+		require.Error(t, err)
+		require.Contains(t, err.Error(), RuleErrorValidatorMissingVotingPublicKeySignature)
+	}
+	{
+		// RuleErrorValidatorInvalidVotingPublicKeySignature: invalid TransactorPkBytes
+		votingPublicKey, votingSignature := _generateVotingPublicKeyAndSignature(t, m1PkBytes, blockHeight)
+		registerMetadata = &RegisterAsValidatorMetadata{
+			Domains:                    [][]byte{[]byte("https://example.com")},
+			VotingPublicKey:            votingPublicKey,
+			VotingPublicKeySignature:   votingSignature,
+			VotingSignatureBlockHeight: blockHeight,
+		}
+		_, err = _submitRegisterAsValidatorTxn(
+			testMeta, m0Pub, m0Priv, registerMetadata, nil, flushToDB,
+		)
+		require.Error(t, err)
+		require.Contains(t, err.Error(), RuleErrorValidatorInvalidVotingPublicKeySignature)
+	}
+	{
+		// RuleErrorValidatorInvalidVotingPublicKeySignature: invalid VotingPublicKey
+		votingPublicKey, _ := _generateVotingPublicKeyAndSignature(t, m0PkBytes, blockHeight)
+		_, votingSignature := _generateVotingPublicKeyAndSignature(t, m0PkBytes, blockHeight)
+		registerMetadata = &RegisterAsValidatorMetadata{
+			Domains:                    [][]byte{[]byte("https://example.com")},
+			VotingPublicKey:            votingPublicKey,
+			VotingPublicKeySignature:   votingSignature,
+			VotingSignatureBlockHeight: blockHeight,
+		}
+		_, err = _submitRegisterAsValidatorTxn(
+			testMeta, m0Pub, m0Priv, registerMetadata, nil, flushToDB,
+		)
+		require.Error(t, err)
+		require.Contains(t, err.Error(), RuleErrorValidatorInvalidVotingPublicKeySignature)
+	}
+	{
+		// RuleErrorValidatorInvalidVotingPublicKeySignature: invalid VotingSignatureBlockHeight
+		votingPublicKey, votingSignature := _generateVotingPublicKeyAndSignature(t, m0PkBytes, blockHeight)
```

Comment:
Just a question - do I need to include a voting pub key & sig when I'm updating my validator entry or ONLY when registering? if I just want to update my domains, is it okay for me to exclude this?

Reply:
That's correct. You need to provide Domains, Voting Public Key, Voting Signature, and Voting Block Height every time you call Register as Validator regardless if you are creating or updating a ValidatorEntry. It's like a stateless upsert.